### PR TITLE
BREAKING CHANGE: remove support for Python3.6

### PR DESCRIPTION
**Pull request by hakandilek**
This is my contribution for #292 as discussed on #317

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -8,7 +8,7 @@ minversion = 3.10
 envlist =
     flake8
     mypy-{locked,lowest}
-    py{310,39,38,37,36}-{locked,lowest}
+    python{3.10,3.9,3.8,3.7}-{locked,lowest}
```
Comment (jkowalleck) [inline tox.ini:11]: :x:  it must be `py{310,39,38,37}`

File: pyproject.toml [hunk 2]
```diff
@@ -54,12 +53,12 @@ tox = "^3.25.0"
...
 autopep8 = "^1.6.0"
-isort = { version = "^5.10.0", python = ">= 3.6.1" }
+isort = { version = "^5.10.0", python = ">= 3.7" }
 flake8 = "^4.0.1"
-flake8-annotations = {version = "^2.7.0", python = ">= 3.6.2"}
+flake8-annotations = {version = "^2.7.0", python = ">= 3.7"}
 flake8-bugbear = "^22.7.1"
-flake8-isort = { version = "^4.1.2", python = ">= 3.6.1" }
-jsonschema = { version = ">= 4.4.0", python = "> 3.6"}
+flake8-isort = { version = "^4.1.2", python = ">= 3.7" }
```
Comment (jkowalleck) [inline pyproject.toml:60]: :x: the python constraint must be removed entirely

File: pyproject.toml [hunk 3]
```diff
@@ -54,12 +53,12 @@ tox = "^3.25.0"
...
+jsonschema = { version = ">= 4.4.0", python = "> 3.7"}
```
Comment (jkowalleck) [inline pyproject.toml:61]: :x: the python constraint must be removed entirely

File: pyproject.toml [hunk 4]
```diff
@@ -54,12 +53,12 @@ tox = "^3.25.0"
 coverage = "^6.2"
 mypy = ">= 0.920, <= 0.961"
 autopep8 = "^1.6.0"
-isort = { version = "^5.10.0", python = ">= 3.6.1" }
+isort = { version = "^5.10.0", python = ">= 3.7" }
 flake8 = "^4.0.1"
-flake8-annotations = {version = "^2.7.0", python = ">= 3.6.2"}
+flake8-annotations = {version = "^2.7.0", python = ">= 3.7"}
```
Comment (jkowalleck) [inline pyproject.toml:58]: :x: the python constraint must be removed entirely
every py37 is >= 3.6.2

File: pyproject.toml [hunk 5]
```diff
@@ -54,12 +53,12 @@ tox = "^3.25.0"
 coverage = "^6.2"
 mypy = ">= 0.920, <= 0.961"
 autopep8 = "^1.6.0"
-isort = { version = "^5.10.0", python = ">= 3.6.1" }
+isort = { version = "^5.10.0", python = ">= 3.7" }
```
Comment (jkowalleck) [inline pyproject.toml:56]: :x: the python constraint must be removed entirely

File: pyproject.toml [hunk 6]
```diff
@@ -42,7 +41,7 @@ keywords = [
 
 [tool.poetry.dependencies]
 # ATTENTION: keep `requirements.lowest.txt` file in sync
-python = "^3.6"
+python = "^3.7"
```
Comment (jkowalleck) [inline pyproject.toml:44]: :+1:

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: tox.ini (hunk 1 above)
Comment (hakandilek) [reply tox.ini:11]: Interpreters are not found with the `py{310,39,38,37}`:

```
SKIPPED:  py39-locked: InterpreterNotFound: python3.9
SKIPPED:  py39-lowest: InterpreterNotFound: python3.9
SKIPPED:  py38-locked: InterpreterNotFound: python3.8
SKIPPED:  py38-lowest: InterpreterNotFound: python3.8
SKIPPED:  py37-locked: InterpreterNotFound: python3.7
SKIPPED:  py37-lowest: InterpreterNotFound: python3.7
```
I've changed as suggested here: https://github.com/tox-dev/tox/issues/378#issuecomment-413937359

File: pyproject.toml (hunk 2 above)
Comment (hakandilek) [reply pyproject.toml]: done

File: pyproject.toml (hunk 5 above)
Comment (hakandilek) [reply pyproject.toml]: done

File: pyproject.toml (hunk 3 above)
Comment (hakandilek) [reply pyproject.toml]: done

File: tox.ini (hunk 1 above)
Comment (jkowalleck) [reply tox.ini:11]: did you have python 3.7 to python 3.9 installed then you had tox tunning? 
tox is intended to skip interpreters that were not installed.

Comment (jkowalleck) [reply tox.ini:11]: please just revert it to the previous config. it was all as intended. 
just remove the `py36` from the list .


read https://tox.wiki/en/latest/config.html

File: .github/workflows/poetry.yml [hunk 7]
```diff
@@ -55,7 +55,8 @@ jobs:
             python-version: '3.10'
             toxenv-factor: 'locked'
           - # test with the lowest dependencies
-            python-version: '3.6'
+            os: ubuntu-20.04
```
Comment (jkowalleck) [inline .github/workflows/poetry.yml:58]: :x: remove the `os` - it uis unused in this matrix

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: .github/workflows/poetry.yml [hunk 8]
```diff
@@ -91,13 +92,16 @@ jobs:
...
         toxenv-factor: ['locked']
         include:
+          - # test with py37 ubuntu20
+            os: ubuntu-20.04
+            python-version: '3.7'
+            toxenv-factor: 'locked'
           - # test with the lowest dependencies
-            os: 'ubuntu-latest'
-            python-version: '3.6'
+            os: ubuntu-20.04
```
Comment (jkowalleck) [inline .github/workflows/poetry.yml:103]: why 20.04? is py37 not working in `latest` anymore

File: tox.ini (hunk 1 above)
Comment (hakandilek) [reply tox.ini]: Yes you're right. I've missed the `skip_missing_interpreters` option there.

File: .github/workflows/poetry.yml (hunk 7 above)
Comment (hakandilek) [reply .github/workflows/poetry.yml]: done!

File: .github/workflows/poetry.yml (hunk 8 above)
Comment (hakandilek) [reply .github/workflows/poetry.yml]: Uh, I think I've cherry-picked this wrong from the previous PR. Set to `latest` now.

File: tox.ini [hunk 9]
```diff
@@ -53,3 +53,5 @@ ignore =
     E305
     # ignore `self`, `cls` markers of flake8-annotations>=2.0
     ANN101,ANN102
+    # ignore ANN401 for dynamically typed *args and **kwargs
+    ANN401
```
Comment (jkowalleck) [inline tox.ini:57]: i wonder why this was needed. 
per https://pypi.org/project/flake8-annotations/ it should be disabled already.

:question: where did it pop up?

**Review by jkowalleck — APPROVED** [review]

**Comment by jkowalleck** [comment]
LGTM.

sometimes the CI/CT failed. Maybe it is due to poetry being this unstable tool it is.

File: tox.ini (hunk 9 above)
Comment (hakandilek) [reply tox.ini:57]: You're right. It should be disabled but my `flake8` checks are failing on Python 3.10.6:

```
flake8 run-test: commands[0] | poetry run flake8 cyclonedx/ tests/
cyclonedx/model/service.py:302:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/bom_ref.py:50:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/issue.py:89:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/issue.py:209:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/component.py:147:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/component.py:299:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/component.py:389:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/component.py:1150:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:119:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:179:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:324:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:381:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:441:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:690:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:758:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/vulnerability.py:1087:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:70:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:84:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:250:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:368:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:431:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:526:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:630:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:702:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:764:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:842:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:915:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:992:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:1071:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:1178:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:1257:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/model/__init__.py:1300:22: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/output/serializer/json.py:44:23: ANN401 Dynamically typed expressions (typing.Any) are disallowed
cyclonedx/output/serializer/json.py:45:0: ANN401 Dynamically typed expressions (typing.Any) are disallowed
tests/base.py:76:45: ANN401 Dynamically typed expressions (typing.Any) are disallowed

```

Comment (jkowalleck) [reply tox.ini:57]: thanks for the info.
thanks for taking care of it.

**Comment by hakandilek** [comment]
> @hakandilek would you be so kind and re-generate a working poetry lock? it it did not work on first try, just leave it as is.

I did regenerate it but I suppose the problem is the incompatibility with the `importlib-metadata` on Python 3.7:

```
poetry 1.3.1 requires importlib-metadata<5.0,>=4.4; python_version < "3.10", but you have importlib-metadata 3.4.0 which is incompatible.
keyring 23.11.0 requires importlib-metadata>=4.11.4; python_version < "3.12", but you have importlib-metadata 3.4.0 which is incompatible.

```

**Comment by jkowalleck** [comment]
@madpah could you have a look at this?

**Comment by madpah** [comment]
Hey @hakandilek - thanks for the contribution.

I just cloned your branch and all tests pass locally for me - I am not getting the `importlib-metadata` conflict that we see in CI - so will have to dig a bit deeper.

**Review by madpah — APPROVED** [review]
Looks good - assuming we can get tests to pass on CI.

**Comment by madpah** [comment]
@hakandilek - found the root cause I believe.

Can you update your branch from main (merge our main into your branch) please?

There was [a fix](https://github.com/CycloneDX/cyclonedx-python-lib/commit/7f8c668cf152af554dbc5183f275723cd3d472b2) added to `tox.ini` which is not in your branch which is breaking the tests.

**Comment by hakandilek** [comment]
> Can you update your branch from main (merge our main into your branch) please?

Well, exactly did that but the DCO is failing now. I suppose there were some unsigned commits coming in.

**Comment by jkowalleck** [comment]
the poetry lock files are incompatible to the version of oetry we are running. 

@madpah do we want to bump our env to poetry 1.3? 
Or should we ask @hakandilek to use v1.1 to generate the logs? 

Should we add this constraint to the dev-requirements in our docs/CONTRIBUTING file? 

FYI: it is possible to have multiple versions of poetry installed in parallel 
Achieved this here via [pipx](https://pypi.org/project/pipx/) ala
```sh
pipx install --suffix=@1.2.2 poetry==1.2.2
```

**Comment by madpah** [comment]
Thanks @jkowalleck. Bit tricky to move `poetry` version right now and we're close to getting some cyclical dependencies between features :-(

We're going to have to take this into `4.0.0` branch via another branch I think to get this moving.

Leave with me @hakandilek / @jkowalleck .

**Pull request merged**